src/course-unit/xblock-container-iframe/index.tsx

```diff
@@ -1,57 +1,183 @@
+      if (foundXBlockInfo) {
+        const { name, userPartitionInfo, blockType } = foundXBlockInfo;
+
+        setCurrentXBlockData({
+          category: COURSE_BLOCK_NAMES.component.id,
+          displayName: name,
+          userPartitionInfo,
+          showCorrectness: 'always',
+          blockType,
+          id: currentXBlockId,
+        });
+      }
+    }
+  }, [isConfigureModalOpen, currentXBlockId, courseVerticalChildren]);
+
+  const handleRefreshIframe = useCallback(() => {
+    // TODO: this artificial delay is a temporary solution
+    // to ensure the iframe content is properly refreshed.
+    setTimeout(() => {
+      sendMessageToIframe(messageTypes.refreshXBlock, null);
+    }, 1000);
+  }, [sendMessageToIframe]);
+
+  const handleDuplicateXBlock = useCallback(
+    (xblockData: XBlockDataTypes) => {
+      const duplicateAndNavigate = (blockType: string) => {
+        unitXBlockActions.handleDuplicate(xblockData.id);
+        if ([COMPONENT_TYPES.html, COMPONENT_TYPES.problem, COMPONENT_TYPES.video].includes(blockType)) {
+          navigate(`/course/${courseId}/editor/${blockType}/${xblockData.id}`);
+        }
+        handleRefreshIframe();
+      };
+
+      duplicateAndNavigate(xblockData.blockType);
+    },
+    [unitXBlockActions, courseId, navigate, handleRefreshIframe],
+  );
+
+  const handleRefetchXBlocks = useCallback(() => {
+    // TODO: this artificial delay is a temporary solution
+    // to ensure the iframe content is properly refreshed.
+    setTimeout(() => {
+      dispatch(fetchCourseUnitQuery(blockId));
+    }, 1000);
+  }, [dispatch, blockId]);
+
+  useEffect(() => {
+    const messageHandlers: Record<string, (payload: MessagePayloadTypes) => void> = {
+      [messageTypes.deleteXBlock]: openDeleteModal,
+      [messageTypes.manageXBlockAccess]: () => openConfigureModal(),
+      [messageTypes.copyXBlock]: () => dispatch(copyToClipboard(currentXBlockId)),
+      [messageTypes.duplicateXBlock]: () => handleDuplicateXBlock(currentXBlockData),
+      [messageTypes.refreshXBlockPositions]: handleRefetchXBlocks,
+      [messageTypes.newXBlockEditor]: ({ url }) => navigate(`/course/${courseId}/editor${url}`),
+      [messageTypes.currentXBlockId]: ({ id }) => setCurrentXBlockId(id),
+      [messageTypes.toggleCourseXBlockDropdown]: ({
+        courseXBlockDropdownHeight,
+      }) => setCourseXBlockIframeOffset(courseXBlockDropdownHeight),
+    };
+
+    const handleMessage = (event: MessageEvent) => {
+      const { type, payload } = event.data || {};
+
+      if (type && messageHandlers[type]) {
+        messageHandlers[type](payload);
+      }
+    };
+
+    window.addEventListener('message', handleMessage);
 
-  const iframeUrl = `${getConfig().STUDIO_BASE_URL}/container_embed/${blockId}`;
+    return () => {
+      window.removeEventListener('message', handleMessage);
+    };
+  }, [dispatch, blockId, courseVerticalChildren, currentXBlockId, currentXBlockData]);
```

Comment:
Complex objects like `courseVerticalChildren` and `currentXBlockData` sometimes cause additional runs of the effect, because they're compared by reference and not by value. That may be OK in this case but it's not ideal.

This is also missing dependencies like `handleDuplicateXBlock`, `handleRefetchXBlocks`, etc. which are also changing.

Since `currentXBlockData` is React state, I would personally just fetch it inside `handleDuplicateXBlock` (`const handleDuplicateXBlock = useCallback(() => { /* do something with currentXBlockData */ }, [currentXBlockData]);`) and then you can simplify this so it's just `[messageTypes.duplicateXBlock]: () => handleDuplicateXBlock` and only depends on `handleDuplicateXBlock`, not directly depending on `currentXBlockData`

----------------

But as I mentioned on the other PR, I think it would be _much_ cleaner to pass in the usage key as part of the event payload, and then you don't need `currentXBlock[Id|Data]` at all, and you can simplify a lot of these, as well as removing most of the dependencies for the `useEffect` and `useCallback` hooks, which will improve performance and reduce re-rendering a lot.

Reply:
Yes, thanks. I like the idea of ​​getting the xblock `id` via `payload`. I added this feature + did some refactoring, please take a look.
https://github.com/openedx/frontend-app-authoring/pull/1431/commits/ea4da9f7625c3ca176d5fd8132f0b23a8329886c